enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -4715,7 +4721,279 @@ class AdjointGenerator
+    std::vector<bool> activities(g.signature.size(), false);
+    std::vector<unsigned> actives;
+    std::map<std::string, unsigned> nm2idx;
+    std::map<unsigned, std::string> idx2nm;
+    for (unsigned i = 0; i != g.signature.size(); ++i) {
+      args[i] = call.getArgOperand(i);
+      types[i] = args[i]->getType();
+      idx2nm[i] = g.signature[i].name;
+      nm2idx[g.signature[i].name] = i;
+      if (g.signature[i].type == Real || g.signature[i].type == RealPointer) {
+        activities[i] = !gutils->isConstantValue(args[i]);
+        if (activities[i]) {
+          actives.push_back(i);
+        }
+      }
+    }
+    std::vector<unsigned> cacheArgIndices;
+    std::map<unsigned, std::vector<unsigned>> dependencies;
+    for (const auto &[name, chainrule] : g.chainrules) {
+      unsigned idx = nm2idx[name];
+      for (const auto &gradientop : chainrule) {
+        for (const auto &arg : gradientop.args) {
+          if (nm2idx.find(arg) != nm2idx.end()) {
+            dependencies[idx].push_back(nm2idx[arg]);
+          }
+        }
+      }
+    }
+    for (unsigned i = 0; i != g.signature.size(); ++i) {
+      if (g.signature[i].type == RealPointer) {
+        bool modified =
+            uncacheable_args.find(calledFunc->arg_begin() + i)->second;
+        bool needed =
+            std::any_of(dependencies[i].begin(), dependencies[i].end(),
+                        [&](unsigned j) { return activities[j]; });
+        if (modified && needed) {
+          cacheArgIndices.push_back(i);
+        }
+      }
+    }
+    // 2. do cache
+    Value *cacheValue;
+    Type *cacheType;
+    std::map<unsigned, unsigned> sz, inc;
+    for (const auto &[name, p] : g.cacheKnowledge) {
+      sz[nm2idx[name]] = nm2idx[p.first];
+      inc[nm2idx[name]] = nm2idx[p.second];
+    }
+    if (!cacheArgIndices.empty()) {
+      std::vector<Type *> cacheFieldTypes =
+          std::for_each(cacheArgIndices.begin(), cacheArgIndices.end(),
+                        [&args](unsigned i) { return args[i]->getType(); });
+      cacheType = StructType::get(call.getContext(), cacheFieldTypes);
+      cacheValue = UndefValue::get(cacheType);
+      if (Mode == DerivativeMode::ReverseModeCombined ||
+          Mode == DerivativeMode::ReverseModePrimal) {
+        for (unsigned cacheRank = 0; cacheRank != cacheArgIndices.size();
+             ++cacheRank) {
+          unsigned index = cacheArgIndices[cacheRank];
+          auto toCopy = gutils->getNewFromOriginal(args[index]);
+          auto malloc =
+              CallInst::CreateMalloc(&newCall, size->getType(), scalarType,
+                                     size, args[sz[index]], nullptr, "");
+          auto bitcast = BuilderZ.CreateBitCast(malloc, toCopy->getType());
+          auto stridedMemcpy = getOrInsertMemcpyStrided(
+              *currentModule, PointerType::getUnqual(scalarType), 0, 0);
+          BuilderZ.CreateCall(stridedMemcpy, {bitcast, toCopy, args[sz[index]],
```

Comment:
I'm not sure this is generalizable like this when we get to more advanced data layouts like matricies.

Reply:
I took a look at the GEMV's description and I found that in `lda != m` case, this approach will result in copying `lda × n` many elements instead of `m × n` (which is less efficient). So you are right! I need to implement a new memcpy function to handle submatrix copying